In [1]:
# Import necessary libraries
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Add the parent directory to the system path 
import os
import sys
sys.path.append(os.path.abspath(os.path.join('..')))

In [3]:
# importing modules from scripts 
from scripts.data_processing import *


In [4]:
# Read the .txt file as a CSV, using '|' as the delimiter
df = pd.read_csv('../data/MachineLearningRating_v3.txt', delimiter='|', low_memory=False)


In [5]:
df.head()

UnderwrittenCoverID  PolicyID     TransactionMonth  IsVATRegistered  \
0               145249     12827  2015-03-01 00:00:00             True   
1               145249     12827  2015-05-01 00:00:00             True   
2               145249     12827  2015-07-01 00:00:00             True   
3               145255     12827  2015-05-01 00:00:00             True   
4               145255     12827  2015-07-01 00:00:00             True   

  Citizenship          LegalType Title Language                 Bank  \
0              Close Corporation    Mr  English  First National Bank   
1              Close Corporation    Mr  English  First National Bank   
2              Close Corporation    Mr  English  First National Bank   
3              Close Corporation    Mr  English  First National Bank   
4              Close Corporation    Mr  English  First National Bank   

       AccountType  ...                    ExcessSelected CoverCategory  \
0  Current account  ...             Mobility - Windscreen    Windscreen   
1  Current account  ...             Mobility - Windscreen    Windscreen   
2  Current account  ...             Mobility - Windscreen    Windscreen   
3  Current account  ...  Mobility - Metered Taxis - R2000    Own damage   
4  Current account  ...  Mobility - Metered Taxis - R2000    Own damage   

    CoverType            CoverGroup              Section  \
0  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
1  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
2  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
3  Own Damage  Comprehensive - Taxi  Motor Comprehensive   
4  Own Damage  Comprehensive - Taxi  Motor Comprehensive   

                           Product StatutoryClass StatutoryRiskType  \
0  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
1  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
2  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
3  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
4  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   

   TotalPremium TotalClaims  
0     21.929825         0.0  
1     21.929825         0.0  
2      0.000000         0.0  
3    512.848070         0.0  
4      0.000000         0.0  

[5 rows x 52 columns]

### Exploratory Data Analysis

In [6]:
# shape of the data 
df.shape

(1000098, 52)

There are 1000098 rows and 52 in the data.

In [7]:
# Display basic information about the dataset
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000098 entries, 0 to 1000097
Data columns (total 52 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   UnderwrittenCoverID       1000098 non-null  int64  
 1   PolicyID                  1000098 non-null  int64  
 2   TransactionMonth          1000098 non-null  object 
 3   IsVATRegistered           1000098 non-null  bool   
 4   Citizenship               1000098 non-null  object 
 5   LegalType                 1000098 non-null  object 
 6   Title                     1000098 non-null  object 
 7   Language                  1000098 non-null  object 
 8   Bank                      854137 non-null   object 
 9   AccountType               959866 non-null   object 
 10  MaritalStatus             991839 non-null   object 
 11  Gender                    990562 non-null   object 
 12  Country                   1000098 non-null  object 
 13  Province                  1

In [8]:
# Display summary statistics
print(df.describe())

       UnderwrittenCoverID      PolicyID    PostalCode        mmcode  \
count         1.000098e+06  1.000098e+06  1.000098e+06  9.995460e+05   
mean          1.048175e+05  7.956682e+03  3.020601e+03  5.487770e+07   
std           6.329371e+04  5.290039e+03  2.649854e+03  1.360381e+07   
min           1.000000e+00  1.400000e+01  1.000000e+00  4.041200e+06   
25%           5.514300e+04  4.500000e+03  8.270000e+02  6.005692e+07   
50%           9.408300e+04  7.071000e+03  2.000000e+03  6.005842e+07   
75%           1.391900e+05  1.107700e+04  4.180000e+03  6.005842e+07   
max           3.011750e+05  2.324600e+04  9.870000e+03  6.506535e+07   

       RegistrationYear      Cylinders  cubiccapacity      kilowatts  \
count      1.000098e+06  999546.000000  999546.000000  999546.000000   
mean       2.010225e+03       4.046642    2466.743258      97.207919   
std        3.261391e+00       0.294020     442.800640      19.393256   
min        1.987000e+03       0.000000       0.000000       0.0

In [9]:
# Let's checkout individual columns
print(df.columns)

Index(['UnderwrittenCoverID', 'PolicyID', 'TransactionMonth',
       'IsVATRegistered', 'Citizenship', 'LegalType', 'Title', 'Language',
       'Bank', 'AccountType', 'MaritalStatus', 'Gender', 'Country', 'Province',
       'PostalCode', 'MainCrestaZone', 'SubCrestaZone', 'ItemType', 'mmcode',
       'VehicleType', 'RegistrationYear', 'make', 'Model', 'Cylinders',
       'cubiccapacity', 'kilowatts', 'bodytype', 'NumberOfDoors',
       'VehicleIntroDate', 'CustomValueEstimate', 'AlarmImmobiliser',
       'TrackingDevice', 'CapitalOutstanding', 'NewVehicle', 'WrittenOff',
       'Rebuilt', 'Converted', 'CrossBorder', 'NumberOfVehiclesInFleet',
       'SumInsured', 'TermFrequency', 'CalculatedPremiumPerTerm',
       'ExcessSelected', 'CoverCategory', 'CoverType', 'CoverGroup', 'Section',
       'Product', 'StatutoryClass', 'StatutoryRiskType', 'TotalPremium',
       'TotalClaims'],
      dtype='object')


In [10]:
# Data Structure of each column 
df.dtypes

UnderwrittenCoverID           int64
PolicyID                      int64
TransactionMonth             object
IsVATRegistered                bool
Citizenship                  object
LegalType                    object
Title                        object
Language                     object
Bank                         object
AccountType                  object
MaritalStatus                object
Gender                       object
Country                      object
Province                     object
PostalCode                    int64
MainCrestaZone               object
SubCrestaZone                object
ItemType                     object
mmcode                      float64
VehicleType                  object
RegistrationYear              int64
make                         object
Model                        object
Cylinders                   float64
cubiccapacity               float64
kilowatts                   float64
bodytype                     object
NumberOfDoors               

### Data Quality Assessment

In [11]:
# Check for missing values
print(df.isnull().sum())

UnderwrittenCoverID               0
PolicyID                          0
TransactionMonth                  0
IsVATRegistered                   0
Citizenship                       0
LegalType                         0
Title                             0
Language                          0
Bank                         145961
AccountType                   40232
MaritalStatus                  8259
Gender                         9536
Country                           0
Province                          0
PostalCode                        0
MainCrestaZone                    0
SubCrestaZone                     0
ItemType                          0
mmcode                          552
VehicleType                     552
RegistrationYear                  0
make                            552
Model                           552
Cylinders                       552
cubiccapacity                   552
kilowatts                       552
bodytype                        552
NumberOfDoors               

In [12]:
# Check for duplicates
print(f"Number of duplicate rows: {df.duplicated().sum()}")

Number of duplicate rows: 0


In [13]:
# percentage of missing values 
calculate_missing_percentage(df)

The dataset has 9.74% missing values.


The below table presents a summary of missing data for various columns in a dataset.

In [14]:
# Data Quality Assessment. # total missing values and it's percentage 
missing_data = check_missing_values(df)
missing_data

Missing Values  % of Total Values Data type
NumberOfVehiclesInFleet          1000098             100.00   float64
CrossBorder                       999400              99.93    object
CustomValueEstimate               779642              77.96   float64
Rebuilt                           641901              64.18    object
Converted                         641901              64.18    object
WrittenOff                        641901              64.18    object
NewVehicle                        153295              15.33    object
Bank                              145961              14.59    object
AccountType                        40232               4.02    object
Gender                              9536               0.95    object
MaritalStatus                       8259               0.83    object
VehicleType                          552               0.06    object
make                                 552               0.06    object
mmcode                               552               0.06   float64
Model                                552               0.06    object
Cylinders                            552               0.06   float64
bodytype                             552               0.06    object
kilowatts                            552               0.06   float64
NumberOfDoors                        552               0.06   float64
VehicleIntroDate                     552               0.06    object
cubiccapacity                        552               0.06   float64
CapitalOutstanding                     2               0.00    object
SubCrestaZone                          0               0.00    object
ItemType                               0               0.00    object
Province                               0               0.00    object
Country                                0               0.00    object
PostalCode                             0               0.00     int64
MainCrestaZone                         0               0.00    object
Title                                  0               0.00    object
Language                               0               0.00    object
TransactionMonth                       0               0.00    object
IsVATRegistered                        0               0.00      bool
UnderwrittenCoverID                    0               0.00     int64
PolicyID                               0               0.00     int64
LegalType                              0               0.00    object
Citizenship                            0               0.00    object
AlarmImmobiliser                       0               0.00    object
TrackingDevice                         0               0.00    object
RegistrationYear                       0               0.00     int64
SumInsured                             0               0.00   float64
TermFrequency                          0               0.00    object
CalculatedPremiumPerTerm               0               0.00   float64
ExcessSelected                         0               0.00    object
CoverCategory                          0               0.00    object
CoverType                              0               0.00    object
CoverGroup                             0               0.00    object
Section                                0               0.00    object
Product                                0               0.00    object
StatutoryClass                         0               0.00    object
StatutoryRiskType                      0               0.00    object
TotalPremium                           0               0.00   float64
TotalClaims                            0               0.00   float64

The dataframe contains 52 columns.
22 columns have missing values.

### Handling Missing values 

drop columns with high percentage of missing values (>50%): 
- NumberOfVehiclesInFleet(100% missing)
- CrossBorder (99.93 %)
- CustomValueEstimate(77.96%)
- Rebuilt(64.18%)
- Converted(64.18%)
- WrittenOff(64.18%)

In [15]:
# Drop columns with missing values above the specified threshold(50%).
df_cleaned= drop_high_missing_columns(df)

Dropped columns: ['CustomValueEstimate', 'WrittenOff', 'Rebuilt', 'Converted', 'CrossBorder', 'NumberOfVehiclesInFleet']


for the rest of the data
- For categorical columns, impute missing values using the mode (most frequent value). 
- For numerical columns, it uses the median to fill in missing values.

In [16]:
df= impute_missing_values(df_cleaned)

check remaining missing values

In [19]:
# Data Quality Assessment. # total missing values and it's percentage 
missing_data = check_missing_values(df)
missing_data

Missing Values  % of Total Values Data type
UnderwrittenCoverID                    0                0.0     int64
PolicyID                               0                0.0     int64
TransactionMonth                       0                0.0    object
IsVATRegistered                        0                0.0      bool
Citizenship                            0                0.0    object
LegalType                              0                0.0    object
Title                                  0                0.0    object
Language                               0                0.0    object
Bank                                   0                0.0    object
AccountType                            0                0.0    object
MaritalStatus                          0                0.0    object
Gender                                 0                0.0    object
Country                                0                0.0    object
Province                               0                0.0    object
PostalCode                             0                0.0     int64
MainCrestaZone                         0                0.0    object
SubCrestaZone                          0                0.0    object
ItemType                               0                0.0    object
mmcode                                 0                0.0   float64
VehicleType                            0                0.0    object
RegistrationYear                       0                0.0     int64
make                                   0                0.0    object
Model                                  0                0.0    object
Cylinders                              0                0.0   float64
cubiccapacity                          0                0.0   float64
kilowatts                              0                0.0   float64
bodytype                               0                0.0    object
NumberOfDoors                          0                0.0   float64
VehicleIntroDate                       0                0.0    object
AlarmImmobiliser                       0                0.0    object
TrackingDevice                         0                0.0    object
CapitalOutstanding                     0                0.0    object
NewVehicle                             0                0.0    object
SumInsured                             0                0.0   float64
TermFrequency                          0                0.0    object
CalculatedPremiumPerTerm               0                0.0   float64
ExcessSelected                         0                0.0    object
CoverCategory                          0                0.0    object
CoverType                              0                0.0    object
CoverGroup                             0                0.0    object
Section                                0                0.0    object
Product                                0                0.0    object
StatutoryClass                         0                0.0    object
StatutoryRiskType                      0                0.0    object
TotalPremium                           0                0.0   float64
TotalClaims                            0                0.0   float64

## Univariate Analysis

In [20]:
# Univariate Analysis for numerical columns 
#numeric_cols = df.select_dtypes(include=[np.number]).columns
#for col in numeric_cols:
 #   plt.figure(figsize=(10, 6))
  #  sns.histplot(df[col], kde=True)
   # plt.title(f'Distribution of {col}')
    #plt.show()